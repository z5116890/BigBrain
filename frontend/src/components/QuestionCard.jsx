import React from 'react';
import {
  makeStyles,
  Card,
  CardContent,
  Typography,
  Button,
} from '@material-ui/core';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import useMediaQuery from '@material-ui/core/useMediaQuery';
import DeleteIcon from '@material-ui/icons/Delete';

// Question card is the individual questions that appear on editsidebar
function QuestionCard(props) {
  const {
    url,
    currQno,
    qno,
    question,
    onClick,
    handleDelete,
  } = props;

  const useStyles = makeStyles({
    root: {
      width: '100%',
      height: '140px',
      margin: 2,
      '&:hover': {
        cursor: 'pointer',
      },
      backgroundColor: (currQno === qno) ? '#ebf4fc' : '#FFFFFF',
    },
    container: {
      width: useMediaQuery('(max-width:700px)') ? '160px' : '100%',
      height: '140px',
      position: 'relative',
      marginTop: '2%',
    },
    title: {
      fontSize: 14,
    },
    pos: {
      marginBottom: 12,
    },
    close: {
      position: 'absolute',
      top: '2px',
      right: '0px',
      '&:hover': {
        cursor: 'pointer',
      },
      width: '15px',
      fontSize: 'large',
    },
  });
  const classes = useStyles();

  return (
    <div id="question-card" className={classes.container}>
      {/* Clickable card */}
      <Button
        aria-label="delete-question-button"
        className={classes.close}
        onClick={(e) => handleDelete(e, qno)}
        color="secondary"
      >
        <DeleteIcon />
      </Button>
      <Link to={`${url}/${qno}`}>
        <Card aria-label={`question-card-${qno}`} id="question-card" hoverable="true" key={qno} onClick={(e) => onClick(e, qno)} className={classes.root}>
          <CardContent>
            <Typography data-test="qno" className={classes.title} color="textSecondary" gutterBottom>
              {`Question ${qno}`}
            </Typography>
            <Typography data-test="question" className={classes.pos} color="textSecondary">
              {question}
            </Typography>
          </CardContent>
        </Card>
      </Link>
    </div>
  );
}
QuestionCard.propTypes = {
  url: PropTypes.string.isRequired,
  currQno: PropTypes.number.isRequired,
  qno: PropTypes.number.isRequired,
  question: PropTypes.string.isRequired,
  onClick: PropTypes.func.isRequired,
  handleDelete: PropTypes.func.isRequired,
};
export default QuestionCard;
