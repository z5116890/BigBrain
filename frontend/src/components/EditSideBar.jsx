import React from 'react';
import { Button } from '@material-ui/core';
import PropTypes from 'prop-types';
import styles from '../css/editGame.module.css';
import QuestionCard from './QuestionCard';

// sidebar contains questions that the user can select to edit the question
// user can also delete or add a question on the sidebar
function EditSideBar(props) {
  const {
    url,
    currQno,
    quizQuestions,
    handleCardClick,
    addNewQuestion,
    handleDelete,
  } = props;
  return (
    <div className={styles.side_container}>
      {/* Create Clickable Card for each question */}
      <div className={styles.questions_list_container}>
        {quizQuestions && quizQuestions.map((q) => <QuestionCard url={url} quizID={q.quizID} currQno={currQno} key={q.qno} qno={q.qno} question={q.question} handleDelete={handleDelete} onClick={handleCardClick}>{' '}</QuestionCard>)}
      </div>
      <div className={styles.add_button}>
        <Button
          aria-label="add-question"
          className={styles.edit_button}
          variant="contained"
          color="primary"
          onClick={addNewQuestion}
        >
          +
        </Button>
      </div>
    </div>
  );
}
EditSideBar.propTypes = {
  url: PropTypes.string.isRequired,
  currQno: PropTypes.number.isRequired,
  quizQuestions: PropTypes.arrayOf(PropTypes.object).isRequired,
  handleCardClick: PropTypes.func.isRequired,
  addNewQuestion: PropTypes.func.isRequired,
  handleDelete: PropTypes.func.isRequired,
};
export default EditSideBar;
