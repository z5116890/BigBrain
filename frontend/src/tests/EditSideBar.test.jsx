import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import { Button, Card } from '@material-ui/core';
import QuestionCard from '../components/QuestionCard';
import EditSideBar from '../components/EditSideBar';

const questions = { quizID: 1, qno: 1, question: 'why?' };
describe('EditSideBar', () => {
  it('should have 1 question card when editsidebar is passed in 1 question', () => {
    const onClick = jest.fn();
    const wrapper = shallow(<EditSideBar url="www" handleDelete={onClick} addNewQuestion={onClick} currQno={1} handleCardClick={onClick} quizQuestions={[questions]} />);
    expect(wrapper.find(QuestionCard)).toHaveLength(1);
  });
  it('has a questioncard with custom question number and question', () => {
    const onClick = jest.fn();
    const wrapper = shallow(<QuestionCard url="localhost" quizID={123} currQno={1} key={1} qno={1} question="why?" handleDelete={onClick} onClick={onClick} />);
    const qno = wrapper.find('[data-test="qno"]');
    expect(qno.text()).toBe('Question 1');
    const question = wrapper.find('[data-test="question"]');
    expect(question.text()).toBe('why?');
  });
  it('has an add button that should trigger when clicked and also has aria-label', () => {
    // test click and aria label
    const onClick = jest.fn();
    const wrapper = shallow(
      <Button
        aria-label="add-question"
        onClick={onClick}
      >
        +
      </Button>,
    );
    wrapper.simulate('click');
    expect(onClick).toHaveBeenCalledTimes(1);
    expect(wrapper.props()['aria-label']).toEqual('add-question');
  });
  it('has a question cards that can be clicked', () => {
    const onClick = jest.fn();
    const wrapper = shallow(<QuestionCard url="localhost" quizID={123} currQno={1} key={1} qno={1} question="Why?" handleDelete={onClick} onClick={onClick} />);
    const card = wrapper.find(Card);
    card.simulate('click');
    expect(onClick).toHaveBeenCalledTimes(1);
  });
  it('has a close button that can be clicked', () => {
    const onClick = jest.fn();
    const wrapper = shallow(<QuestionCard url="localhost" quizID={123} currQno={1} key={1} qno={1} question="Why?" handleDelete={onClick} onClick={onClick} />);
    const card = wrapper.find(Button);
    card.simulate('click');
    expect(onClick).toHaveBeenCalledTimes(1);
  });
  // Snapshots. Making sure component does not change unexpectedly by
  // comparing a reference snapshot with test snapshot.
  it('renders with props', () => {
    const onClick = jest.fn();
    const esb = renderer.create(<EditSideBar url="www" handleDelete={onClick} addNewQuestion={onClick} currQno={1} handleCardClick={onClick} quizQuestions={[]} />);
    expect(esb).toMatchSnapshot();
  });
});
