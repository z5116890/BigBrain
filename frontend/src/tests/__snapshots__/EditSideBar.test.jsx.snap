// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditSideBar renders with props 1`] = `
<div
  className="side_container"
>
  <div
    className="questions_list_container"
  />
  <div
    className="add_button"
  >
    <button
      aria-label="add-question"
      className="MuiButtonBase-root MuiButton-root MuiButton-contained edit_button MuiButton-containedPrimary"
      disabled={false}
      onBlur={[Function]}
      onClick={[MockFunction]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        +
      </span>
    </button>
  </div>
</div>
`;
